### PR TITLE
Pull game directory from op2ext

### PR DESCRIPTION
**Pull request by Brett208**
This allows attaching a debugger and still finding the correct location of the .ini file. Otherwise if the debugger is attached, the default protocol index of 4 is returned (because .ini file is not found) which makes it appear as if the module is not loading properly. Guess this whole problem may not exist depending on how we finalize setting protocolIndex. Although we are pulling other values from the .ini file, so this piece of code should be relevant in other parts of the project.

I wanted to find a solution that would allow Outpost 2 to be buried in a directory deeper than MAX_PATH, but didn't come up with anything useful. 

I'm okay in this form or continuing to refactor.

### CODE REVIEW

File: Main.cpp [hunk 1]
```diff
@@ -57,10 +62,24 @@ extern "C" __declspec(dllexport) void InitMod(char* iniSectionName)
...
 	strncpy(sectionName, iniSectionName, sizeof(sectionName));
 
-	int protocolIndex;
-	// Get button index to replace functionality of
-	protocolIndex = GetPrivateProfileInt(sectionName, "ProtocolIndex", DefaultProtocolIndex, ".\\Outpost2.ini");
+
+	// Get multiplayer button index that NetFix will replace
+	const std::string iniPath = GetOutpost2Directory() + "Outpost2.ini"; 
+
+	int protocolIndex = GetPrivateProfileInt(sectionName, "ProtocolIndex", DefaultProtocolIndex, iniPath.c_str());
```
Comment (DanRStevens) [inline Main.cpp:69]: Maybe no blank line separating this.

**Review by DanRStevens — COMMENTED** [review]
AppVeryor failed. The linker is having trouble finding `GetGameDir_s`. I don't see the reason why.

----

In terms of supporting longer paths, you can call `GetGameDir_s` twice, once with a zero sized buffer, then use the returned value to pre-allocate a buffer, then make a second call with the updated size to write into it. If you use a `std::string` as the pre-allocated buffer, you can avoid a copy to it later.

At least that's the double call standard idiom the Windows API uses for allocating and filling buffers of unknown size, where the return value is how big the buffer needed to be.

----

This code has got me wondering if op2ext should support an optional C++ interface.

----

When setting the debug executable, can you also set the path? Is that info stored in the project file, or a non-versioned user file? My guess is, a non-versioned user file. Otherwise that might be the update to make.

That point aside, I don't see anything wrong with the idea presented here.

----

Edit: Tagging Issue #1 as related.

**Comment by Brett208** [comment]
I think the trouble may be from:

```C++
#ifdef OP2EXT_INTERNAL_BUILD  // Defined in op2ext project settings
#define OP2EXT_API __declspec(dllexport)
#else
#define OP2EXT_API __declspec(dllimport) 
#endif
```

We probably need to undefine OP2EXT_INTERNAL_BUILD when NetHelperClient compiles to indicate it is importing the functions.

---

Correct that you can specify the working directory that the debugger is being used. I think this is a symptom of an underlying problem where we really want the game directory and not the environment's working directory to grab the .ini file though. Debugger settings in VS are contained in user files so they do not typically make it into the repository.

Maybe you could unpack the statement about a C++ interface on op2ext. I thought we had to go with a C interface if we wanted to support different compilers since each compiler had its interpretation of the external portions of C++. Maybe even changing between versions of the same compiler. Maybe better to open a branch to not hijack this work.

-Brett

**Comment by DanRStevens** [comment]
Interesting observation about the `#define`. Looking into it though, I don't think that is the cause. The define is only in the op2ext project settings. When the header is included from NetFixClient, that define won't be set, and so the function should be properly imported.

I don't believe the `#define` project setting from op2ext would be imported by NetFixClient. I tried removing the reference from NetFixClient to op2ext, and the error still persisted.

A couple of side notes from the above experiment. The op2ext project was still built by the NetFixClient solution, just it got built after the NetFixClient project once the reference was removed. The op2ext project does produce a static library (.lib) file, which likely contains stubs for the exported functions. I suspect it's use is optional when using MSVC with `__declspec(import)` being used.

**Comment by DanRStevens** [comment]
Maybe read:
https://stackoverflow.com/questions/1380829/is-extern-c-only-required-on-the-function-declaration

It looks like we are including op2ext.h properly from op2ext.cpp. Still though, maybe try adding `extern "C"` to the function definition in op2ext.cpp just to test. It shouldn't need to be there, assuming proper includes, but perhaps something is going wrong and the compiler isn't seeing the `extern "C"` definition, and so is exporting a mangled name, hence a link error when it tries to resolve the name.

Note: This is a test, not so much a suggestion for a permanent fix. If this reveals something, it might lead us to a problem elsewhere.

**Comment by Brett208** [comment]
Need to address concerns in op2ext before merging even if we are happy with everything else.

**Review by DanRStevens — COMMENTED** [review]
This update looks like it's doing the right thing. We just have to solve the link problem to actually test it.

I pushed an update to op2ext for the dependent change. That also needs testing. I'm unable to compile or test op2ext from Linux.

**Comment by DanRStevens** [comment]
Ahh,  now we have a conflict between the `Log(const char* message)` method in NetfixClient, and the `Log(const char* message)` method from op2ext.

**Comment by Brett208** [comment]
I think this should be merged now. We will eventually want to fix the name collision, but that should probably be addressed in op2ext. We may want to update how we reference op2ext from NetHelper as well, but it would be cleaner to address that in a different branch.

-Brett

**Comment by DanRStevens** [comment]
I think this is ready for merging.

The immediate problem of the name collision has been resolved. We can ignore it for now, as long as we don't import both `Log.h` and `op2ext.h` in the same compilation unit.

Long term, the local logging code is likely to go, with the format methods moved elsewhere, so there shouldn't be a name conflict in the future.

**Pull request merged**